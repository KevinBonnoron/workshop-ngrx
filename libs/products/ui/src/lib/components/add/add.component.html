<h1>Ajout d'un produit</h1>

<products-ui-form (save)="onSave($event)"></products-ui-form>
