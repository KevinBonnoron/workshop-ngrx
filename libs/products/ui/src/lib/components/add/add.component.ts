import { Component, inject } from '@angular/core';
import { Router } from '@angular/router';
import { ProductsService } from '@workshop/products-data-access';
import { Product } from '@workshop/shared';
import { FormComponent } from '../form/form.component';

@Component({
  standalone: true,
  imports: [FormComponent],
  templateUrl: './add.component.html',
})
export class AddComponent {
  private readonly router = inject(Router);
  private readonly productsService = inject(ProductsService);

  onSave({ name }: Omit<Product, 'id'>) {
    this.productsService.create({ name }).subscribe(() => this.router.navigateByUrl('list'));
  }
}
