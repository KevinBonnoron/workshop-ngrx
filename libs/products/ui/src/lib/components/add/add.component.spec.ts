import { ComponentFixture, TestBed } from '@angular/core/testing';
import { Router } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';
import { ProductsService } from '@workshop/products-data-access';
import { Product } from '@workshop/shared';
import { of } from 'rxjs';
import { AddComponent } from './add.component';

class BlankComponent {}

describe('AddComponent', () => {
  let fixture: ComponentFixture<AddComponent>;
  let component: AddComponent;
  let productsService: ProductsService;
  let router: Router;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule.withRoutes([{ path: 'list', component: BlankComponent }]),
        AddComponent
      ],
      providers: [
        { provide: ProductsService, useValue: { create: jest.fn((product: Product) => of(product)) } }
      ]
    });

    fixture = TestBed.createComponent(AddComponent);
    component = fixture.componentInstance;
    productsService = TestBed.inject(ProductsService);
    router = TestBed.inject(Router);
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should call endpoint when onSave is called', () => {
    const productServiceCreateSpy = jest.spyOn(productsService, 'create');
    const routerNavigateByUrlSpy = jest.spyOn(router, 'navigateByUrl').mockImplementation(() => Promise.resolve(true));

    const product: Omit<Product, 'id'> = { name: 'Test' };
    component.onSave(product);

    expect(productServiceCreateSpy).toHaveBeenCalledWith(product);
    expect(routerNavigateByUrlSpy).toHaveBeenCalledWith('list');
  });

  it('should match snaptshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
