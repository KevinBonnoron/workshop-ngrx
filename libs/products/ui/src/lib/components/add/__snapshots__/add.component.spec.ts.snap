// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddComponent should match snaptshot 1`] = `
<ng-component
  productsService={[Function Object]}
  router={[Function _Router]}
>
  <h1>
    Ajout d'un produit
  </h1><products-ui-form>
    <form
      novalidate=""
    >
      <mat-form-field
        appearance="outline"
        class="mat-mdc-form-field"
      >
        <div
          class="mat-mdc-text-field-wrapper mdc-text-field"
        >
          <div
            class="mat-mdc-form-field-flex"
          >
            <div
              class="mat-mdc-form-field-infix"
            >
              <input
                class="mat-mdc-input-element"
                formcontrolname="name"
                matinput=""
              />
            </div>
          </div>
        </div>
        <div
          class="mat-mdc-form-field-subscript-wrapper mat-mdc-form-field-bottom-align"
        >
        </div>
      </mat-form-field>
      <button
        class="mdc-button mdc-button--raised mat-mdc-raised-button mat-primary"
        color="primary"
        mat-raised-button=""
        mat-ripple-loader-class-name="mat-mdc-button-ripple"
        mat-ripple-loader-uninitialized=""
      >
        <span
          class="mat-mdc-button-persistent-ripple"
        />
        <span
          class="mdc-button__label"
        >
          Valider
        </span>
        <span
          class="mat-mdc-focus-indicator"
        />
        <span
          class="mat-mdc-button-touch-target"
        />
      </button>
    </form>
  </products-ui-form>
</ng-component>
`;
